### PR TITLE
feat(cfw): add resource CFW dns resolution 

### PR DESCRIPTION
**Pull request by profoundwu**
**What this PR does / why we need it**:
add cfw dns resolution resource
**Which issue this PR fixes**:
*(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*
fixes #xxx

**Special notes for your reviewer**:

**Release note**:

```release-note
1. add cfw dns resolution resource
```

## PR Checklist

* [x] Tests added/passed.
* [x] Documentation updated.
* [x] Schema updated.

## Acceptance Steps Performed

```
make testacc TEST="./huaweicloud/services/acceptance/cfw" TESTARGS="-run TestAccDnsResolution_basic"
==> Checking that code complies with gofmt requirements...
TF_ACC=1 go test ./huaweicloud/services/acceptance/cfw -v -run TestAccDnsResolution_basic -timeout 360m -parallel 4
=== RUN   TestAccDnsResolution_basic
=== PAUSE TestAccDnsResolution_basic
=== CONT  TestAccDnsResolution_basic
--- PASS: TestAccDnsResolution_basic (33.20s)
PASS
ok      github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/services/acceptance/cfw       33.271s

```

### CODE REVIEW

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go [hunk 1]
```diff
@@ -0,0 +1,395 @@
...
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				Description: `The ID of the firewall.`,
+			},
+			"default_dns_servers": {
+				Type: schema.TypeSet,
+				Elem: &schema.Schema{
+					Type: schema.TypeString,
+					ValidateFunc: validation.StringMatch(
```
Comment (houpeng80) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go:54]: delete all ValidateFunc

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go [hunk 2]
```diff
@@ -0,0 +1,395 @@
...
+	differenceServers := difference(userDefaultServers, defaultDNSServers)
+
+	return map[string]interface{}{
+		"dns_server":               buildDNSServersBodyParams(userDefaultServers, differenceServers, userCustomServers),
+		"health_check_domain_name": utils.ValueIngoreEmpty(d.Get("health_check_domain_name")),
+	}
+}
+
+func difference(userDefaultServers, defaultDNSServers []interface{}) []interface{} {
+	userDefaultServersSet := make(map[string]bool)
```
Comment (houpeng80) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]:
```suggestion
	userDefaultServersSet := make(map[string]struct{})
```
change it to struct is more effective

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go [hunk 3]
```diff
@@ -0,0 +1,395 @@
...
+	"github.com/huaweicloud/terraform-provider-huaweicloud/huaweicloud/utils"
+)
+
+// @API CFW GET /v1/{project_id}/dns/servers
+// @API CFW PUT /v1/{project_id}/dns/servers
+func ResourceDNSResolution() *schema.Resource {
+	return &schema.Resource{
+		CreateContext: resourceDNSResolutionCreate,
+		ReadContext:   resourceDNSResolutionRead,
+		UpdateContext: resourceDNSResolutionUpdate,
```
Comment (houpeng80) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]: the update method is same as the create method, just keep one

File: huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_dns_resolution_test.go [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+		product = "cfw"
+	)
+	client, err := cfg.NewServiceClient(product, region)
+	if err != nil {
+		return nil, fmt.Errorf("error creating CFW client: %s", err)
+	}
+
+	path := client.Endpoint + httpUrl
+	path = strings.ReplaceAll(path, "{project_id}", client.ProjectID)
+	path += "?offset=0&limit=50"
```
Comment (houpeng80) [inline huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_dns_resolution_test.go]:
```suggestion
	path += "?offset=0&limit=100"
```
keep the same with the read method of the resource

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go (hunk 1 above)
Comment (profoundwu) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go:54]: fixed.

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go (hunk 2 above)
Comment (profoundwu) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]: fixed

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go (hunk 3 above)
Comment (profoundwu) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]: fixed

File: huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_dns_resolution_test.go (hunk 4 above)
Comment (profoundwu) [reply huaweicloud/services/acceptance/cfw/resource_huaweicloud_cfw_dns_resolution_test.go]: fixed

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+				Type:        schema.TypeString,
+				Optional:    true,
+				Computed:    true,
+				Description: `The health check domain name.`,
+			},
+		},
+	}
+}
+
+func resourceDNSResolutionCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]:
```suggestion
func resourceDNSResolutionCreateOrUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
```

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+	userDefaultDNSServers := d.Get("default_dns_servers").(*schema.Set).List()
+	constains := utils.StrSliceContainsAnother(utils.ExpandToStringList(defaultDNSServers), utils.ExpandToStringList(userDefaultDNSServers))
+	if !constains {
+		return diag.Errorf("error creating CFW DNS resolution configuration: some non-default DNS servers have been specified")
+	}
+
+	userCustomServers := d.Get("custom_dns_servers").(*schema.Set).List()
+	if len(userCustomServers) == 0 && len(userDefaultDNSServers) == 0 {
+		return diag.Errorf("error creating CFW DNS resolution configuration: no DNS servers have been specified")
+	}
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]: we can put this validate in schema

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+				Required:    true,
+				ForceNew:    true,
+				Description: `The ID of the firewall.`,
+			},
+			"default_dns_servers": {
+				Type:        schema.TypeSet,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Optional:    true,
+				Description: `The default DNS servers.`,
+			},
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]:
```suggestion
			"default_dns_servers": {
				Type:        schema.TypeSet,
				Elem:        &schema.Schema{Type: schema.TypeString},
				Optional:    true,
				Description: `The default DNS servers.`,
				MinItems:    1,
				AtLeastOneOf: []string{"default_dns_servers", "custom_dns_servers"},
			},
```

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+			"custom_dns_servers": {
+				Type:        schema.TypeSet,
+				Optional:    true,
+				Elem:        &schema.Schema{Type: schema.TypeString},
+				Description: `The custom DNS servers.`,
+			},
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]:
```suggestion
			"custom_dns_servers": {
				Type:        schema.TypeSet,
				Optional:    true,
				Elem:        &schema.Schema{Type: schema.TypeString},
				MinItems:    1,
				Description: `The custom DNS servers.`,
			},
```

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+	if err != nil {
+		return diag.Errorf("error creating CFW DNS resolution configuration: %s", err)
+	}
+
+	_, err = utils.FlattenResponse(resp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(fwInstanceID)
```
Comment (Jason-Zhang9309) [inline huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]:
```suggestion
        if d.IsNewResource() {
	  d.SetId(fwInstanceID)
	}
```

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go (hunk 6 above)
Comment (profoundwu) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]: fixed

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go (hunk 7 above)
Comment (profoundwu) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]: fixed

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go (hunk 9 above)
Comment (profoundwu) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]: fixed

File: huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go (hunk 5 above)
Comment (profoundwu) [reply huaweicloud/services/cfw/resource_huaweicloud_cfw_dns_resolution.go]: fixed

**Pull request merged**